src/evm/protocol/uniswap_v4/state.rs

```diff
@@ -43,6 +67,200 @@ impl UniswapV4State {
                 .expect("tick_spacing should always be positive"),
             ticks,
         );
-        UniswapV4State { liquidity, sqrt_price, lp_fee, protocol_fees, tick, ticks: tick_list }
+        UniswapV4State { liquidity, sqrt_price, fees, tick, ticks: tick_list }
+    }
+
+    fn swap(
+        &self,
+        zero_for_one: bool,
+        amount_specified: I256,
+        sqrt_price_limit: Option<U256>,
+    ) -> Result<SwapResults, SimulationError> {
```

Comment:
Sorry I'm not very familiar with USV4.

But why are we implementing a `swap` method with `SwapResults` instead of a `get_amount_out` like all the other pool states? If there is a good reason perhaps we should document that in the code?

Reply:
this is a UniswapV4 specific method. then the implementation of `ProtocolSim` for uni v4 will have the `get_amount_out` that will eventually call this method!

Reply:
`get_amount_out` will be implemented as part of the `ProtocolSim` trait implementation. This function is just an internal function that will be used in `get_amount_out`.

I suppose we could directly have everything inside `get_amount_out`, but maybe it would be too much for a single function? 

For now I think it's alright like this because it's consistent with `UniswapV3`

Reply:
Ok ok nice! You can keep it as is - no worries